RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripMergeZeroSuppression.cc

```diff
@@ -0,0 +1,147 @@
+
+  using rawtoken_t = edm::EDGetTokenT<edm::DetSetVector<SiStripRawDigi>>;
+  using zstoken_t = edm::EDGetTokenT<edm::DetSetVector<SiStripDigi>>;
+  rawtoken_t m_rawDigisToMerge;
+  zstoken_t m_zsDigisToMerge;
+};
+
+
+SiStripMergeZeroSuppression::SiStripMergeZeroSuppression(const edm::ParameterSet& conf)
+  : m_algorithms(SiStripRawProcessingFactory::create(conf.getParameter<edm::ParameterSet>("Algorithms")))
+{
+  m_rawDigisToMerge = consumes<edm::DetSetVector<SiStripRawDigi>>(conf.getParameter<edm::InputTag>("DigisToMergeVR"));
+  m_zsDigisToMerge = consumes<edm::DetSetVector<SiStripDigi>>(conf.getParameter<edm::InputTag>("DigisToMergeZS"));
+  produces<edm::DetSetVector<SiStripDigi>>("ZeroSuppressed");
+}
+
+void SiStripMergeZeroSuppression::produce(edm::Event& event, const edm::EventSetup& /*eventSetup*/)
+{
+  LogTrace("SiStripMergeZeroSuppression::produce") << "Starting merging " << "\n";
+  edm::Handle<edm::DetSetVector<SiStripDigi>> inputdigi;
+  edm::Handle<edm::DetSetVector<SiStripRawDigi>> inputraw;
+  event.getByToken(m_rawDigisToMerge, inputdigi);
+  event.getByToken(m_zsDigisToMerge, inputraw);
+
+  LogTrace("SiStripMergeZeroSuppression::produce") << inputdigi->size() << " " << inputraw->size() << "\n";
+  if ( ! inputraw->empty() ) {
+    std::vector<edm::DetSet<SiStripDigi>> outputdigi(inputdigi->begin(), inputdigi->end());
+
+    LogTrace("SiStripMergeZeroSuppression::produce") << "Looping over the raw data collection " << "\n";
+    for ( const auto& rawDigis : *inputraw ) {
+      edm::DetSet<SiStripRawDigi>::const_iterator itRawDigis = rawDigis.begin();
+      uint16_t nAPV = rawDigis.size()/128;
+      uint32_t rawDetId = rawDigis.id;
+
+      std::vector<bool> restoredAPV(std::size_t(nAPV), false);
+
+      bool isModuleRestored = false;
+      for ( uint16_t strip = 0; strip < rawDigis.size(); ++strip ) {
+        if(itRawDigis[strip].adc()!=0){
+          restoredAPV[strip/128] = true;
+          isModuleRestored = true;
+        }
+      }
+
+      if ( isModuleRestored ) {
+        LogTrace("SiStripMergeZeroSuppression::produce") << "Apply the ZS to the raw data collection " << "\n";
+        edm::DetSet<SiStripDigi> suppressedDigis(rawDetId);
+        m_algorithms->suppressVirginRawData(rawDigis, suppressedDigis);
+
+        if ( ! suppressedDigis.empty() ) {
+          LogTrace("SiStripMergeZeroSuppression::produce") << "Looking for the detId with the new ZS in the collection of the zero suppressed data" << "\n"; 
+          bool isModuleInZscollection = false;
+          uint32_t zsDetId{0};
+          auto zsModule = std::lower_bound(std::begin(outputdigi), std::end(outputdigi), rawDetId,
+              [] ( const edm::DetSet<SiStripDigi>& elem, uint32_t testDetId ) { return elem.id < testDetId; });
+          if ( ( std::end(outputdigi) != zsModule ) && ( zsModule->id == rawDetId ) ) {
+            isModuleInZscollection = true;
+            zsDetId = zsModule->id;
```

Comment:
this is flagged by the static analyzer https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24339/30196/llvm-analysis/report-044770.html#EndPath
please check, esp to confirm that this is not a bug

Reply:
Technically `zsDetId` is used, but only in the `LogTrace` below (which I suppose the static analyzer doesn't see after preprocessing). I moved `zsModule->id` into the `LogTrace` line such that no local variable is needed, that should silence the static analyzer warning.

Reply:
you are right.
@smuzaffar @Dr15Jones I recall that at some point the static analyzer was supposed to compile code with EDM_ML_DEBUG enabled. I seem to recall to even have to fix some parts to make them compile for the static analyzer in #21795